### DISCUSSION TITLE
How do I load a yolov5 module and make inferences?

### DISCUSSION THREAD

**Discussion by dstampher** [xamarin/TorchSharp#377] [General]
There is pretty much 0 documentation for this repo.  I have no idea how to use it.  I tried to google this repo and didn't find anything helpful that answers my question.

**Comment by GeorgeS2019** [comment]
I did a quick search for pytorch and yolov5 => [YOLOv5 🚀 in PyTorch](https://github.com/ultralytics/yolov5)

- If you are interested of porting yolov5 module from python to  TorchSharp, the source above could be your first starting point.

 - If you are interested to use Yolov5 pytorch model, you need to create[ a yolov5 model that is compatible to TorchSharp loading](https://github.com/dotnet/TorchSharp/blob/main/docfx/articles/saveload.md)

- However, if you are only interested to use Yolov5 in .NET, here is an alternative option: [ ML.NET sample with yolov5.onnx](https://github.com/mentalstack/yolov5-net)

Question to @NiklasGustafsson and @dsyme => is there a need for TorchSharp to support loading e.g yolov5.onnx

**Comment by NiklasGustafsson** [comment]
@dstampher -- we're working on it, but for a starting point, please see the examples under the src/ folder. 

Right now, the TorchSharp APIs are focused on providing and e2e experience in .NET, starting with data pre-processing, training, evaluation, and then inferencing. That said, there is a way to export PyTorch model weights and loading them in TorchSharp, but you still have to replicate the model code in order for the weights to make any sense. This is described in [saveload.md](https://github.com/dotnet/TorchSharp/blob/main/docfx/articles/saveload.md)

ML.NET, which has product support, offers more in terms of loading existing models for inferencing.

TorchSharp is available on NuGet, including supporting native code redistributable packages: libtorch-XXX (the name depends on what environment you are using -- Windows, Linux, MacOS, and whether you have an Nvidia GPU. You want version 0.93.1 of TorchSharp and 1.9.0.11 of libtorch-XXX.

**Comment by NiklasGustafsson** [comment]
@GeorgeS2019 -- as for loading ONNX models, that's something that we will get to sometime in the medium-term future. I expect that to be post-v1.0.

Given the fact that loading ONNX models for inferencing is already available to .NET users via ML.NET, I expect the support in TorchSharp to first focus on transfer learning scenarios, to the extent that the two are distinct.

**Comment by dstampher** [comment]
@GeorgeS2019 Thanks for linking me to that YOLOv5 C# repo, that might work for my use case.

**Reply by GeorgeS2019** [reply]
@dstampher  TorchSharp has good coverage of [TorchVision](https://github.com/dotnet/TorchSharp/tree/main/src/TorchSharp/TorchVision)  You may need that later when your use case requires more advanced deep learning image analysis application - not yet available in .NET but in pytorch :-)
